libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1230,6 +1246,29 @@ string YulUtilFunctions::resizeDynamicByteArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::cleanUpDynamicByteArrayEndSlotsFunction(ArrayType const& _type)
+{
+	solAssert(_type.isByteArray(), "");
+	solAssert(_type.isDynamicallySized(), "");
+
+	string functionName = "clean_up_byte_array_end_slots_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&](vector<string>& _args, vector<string>&) {
+		_args = {"array", "len", "pos"};
+		return Whiskers(R"(
+			if gt(len, 31) {
+				let dataArea := <dataLocation>(array)
+				let deleteStart := add(dataArea, div(<roundUp>(pos), 32))
```

Comment:
Maybe we should introduce a new helper for rounding and dividing by 32? This is used quite a lot, also in this PR at another place. Can be done in a subsequent PR, though.

Reply:
https://github.com/ethereum/solidity/pull/11271#discussion_r619186073 -- I was kind of sure we had a roundUp32 function.

Reply:
I saw your comment - did you mean `rund up and divide by 32` by "roundUp32"?
Maybe call it `divideBy32Ceil`?

Reply:
```
        /// @returns the name of a function that rounds its input to the next multiple
        /// of 32 or the input if it is a multiple of 32.
        /// signature: (value) -> result
        std::string roundUpFunction();
```

Only meant the "round up to multiple of 32". Probably we should name this helper `roundTo32Function()` or something like that.

As regards to `divideBy32Ceil` -- I was wondering if it is actually prevalent, but it does seem to be used in a fair few places:
```
YulUtilFunctions.cpp:					let deleteStart := add(arrayDataStart, div(add(newLen, 31), 32))
YulUtilFunctions.cpp:					<clearStorageRange>(deleteStart, add(arrayDataStart, div(add(oldLen, 31), 32)))
YulUtilFunctions.cpp:						<clearStorageRange>(add(arrayDataStart, 1), add(arrayDataStart, div(add(oldLen, 31), 32)))
YulUtilFunctions.cpp:					let deleteStart := add(dstDataArea, div(add(newLen, 31), 32))
YulUtilFunctions.cpp:					<clearStorageRange>(deleteStart, add(dstDataArea, div(add(oldLen, 31), 32)))
```